### PR TITLE
Getting relationship DAO from a method on the java side

### LINKED ISSUE #1373: Getting the DAO for a relationship on the java side does not work.
Ex: ``user.getPartners()``
This returns DAO which has the property ``isDelegateSet_ = true`` 
but the delegate is actually ``null`` and hence throws a NullPointerException in ProxyDAO. when a ``select`` is done on that DAO.

In order to make this work you have to do the following:
 ``user.fclone()`` 
 ``user.setX(x)`` 
This returns a correct DAO and the ``select`` works as intended.

### PR DESCRIPTION
**Pull request by shreymahendru**
Fix #1373

### CODE REVIEW

File: src/foam/dao/Relationship.js [hunk 1]
```diff
@@ -129,18 +132,17 @@ foam.CLASS({
     function initRelationship() {
       var sourceProp;
       var targetProp;
-      var cardinality   = this.cardinality;
-      var forwardName   = this.forwardName;
-      var inverseName   = this.inverseName;
-      var relationship  = this;
-      var sourceModel   = this.sourceModel;
-      var targetModel   = this.targetModel;
+      var cardinality = this.cardinality;
```
Comment (kgrgreer) [inline src/foam/dao/Relationship.js]: Why mess up the alignment?

File: src/foam/dao/Relationship.js [hunk 2]
```diff
@@ -365,8 +365,8 @@ foam.CLASS({
       factory: function() {
         return this.__context__[this.junctionDAOKey];
       },
-      javaFactory: 'return (foam.dao.DAO)getX().get(getJunctionDAOKey());',
-      swiftFactory: 'return __context__[junctionDAOKey] as? (foam_dao_DAO & foam_core_FObject)'
+      javaFactory: `return (foam.dao.DAO)getX().get(getJunctionDAOKey());`,
```
Comment (kgrgreer) [inline src/foam/dao/Relationship.js]: Why switch to multi-line string when when not multi-line?

File: src/foam/dao/Relationship.js [hunk 3]
```diff
@@ -375,38 +375,46 @@ foam.CLASS({
       factory: function() {
         return this.__context__[this.targetDAOKey];
       },
-      javaFactory: 'return (foam.dao.DAO)getX().get(getTargetDAOKey());',
-      swiftFactory: 'return __context__[targetDAOKey] as? (foam_dao_DAO & foam_core_FObject)'
+      javaFactory: `return (foam.dao.DAO)getX().get(getTargetDAOKey());`,
```
Comment (kgrgreer) [inline src/foam/dao/Relationship.js]: Why change?

File: src/foam/dao/Relationship.js [hunk 4]
```diff
@@ -375,38 +375,46 @@ foam.CLASS({
...
       name: 'createJunction',
-      args: [ { name: 'targetId', javaType: 'Object' } ],
+      args: [{ name: 'targetId', javaType: 'Object' }],
       returns: 'foam.core.FObject',
       javaReturns: 'foam.core.FObject',
       code: function createJunction(targetId) {
         foam.assert( ( ! foam.Undefined.isInstance(this.sourceId) ) &&
                      ( ! foam.Undefined.isInstance(targetId) ),
-                    "Cannot create an association with an object that isn't stored in a DAO yet.");
+                    `Cannot create an association with an object that isn\'t stored in a DAO yet.`);
```
Comment (kgrgreer) [inline src/foam/dao/Relationship.js]: Why multi-line string for single-line string?

File: src/foam/dao/Relationship.js [hunk 5]
```diff
@@ -428,7 +436,9 @@ return junction`
       returns: 'foam.dao.DAO',
       javaCode: 'return getDao();',
       swiftCode: 'return dao!',
-      code: function getDAO() { return this.dao; }
+      code: function getDAO() {
```
Comment (kgrgreer) [inline src/foam/dao/Relationship.js]: The original seems better.

File: src/foam/dao/Relationship.js [hunk 6]
```diff
@@ -449,10 +459,15 @@ return junction`
         }, x);
 
         controller.sub('select', function(s, _, id) {
-          dao.find(id).then(function(obj) { self.add(obj); });
+          dao.find(id).then(function(obj) {
```
Comment (kgrgreer) [inline src/foam/dao/Relationship.js]: Original seems better. Don't break up to multiple lines if it fits in one reasonable size line.

File: src/foam/dao/Relationship.js [hunk 7]
```diff
@@ -472,10 +487,15 @@ return junction`
         }, x);
 
         controller.sub('select', function(s, _, id) {
-          dao.find(id).then(function(obj) { self.remove(obj); });
+          dao.find(id).then(function(obj) {
```
Comment (kgrgreer) [inline src/foam/dao/Relationship.js]: Original seems better.

File: src/foam/dao/Relationship.js [hunk 8]
```diff
@@ -488,13 +508,14 @@ foam.CLASS({
...
-    ['cloneProperty', function(value, map){}],
+    ['cloneProperty', function(value, map) {}],
     ['javaCloneProperty', '//noop'],
     ['javaDiffProperty', '//noop'],
+    ['generateJava', false],
     {
       class: 'String',
       name: 'targetPropertyName',
-      documentation: "We don't just use targetProperty here because at the time that this axiom is created, the target property may not even be installed yet on the target.  So instead we use a combination of targetPropertyName and target class and get the actual property when needed."
+      documentation: `We don\'t just use targetProperty here because at the time that this axiom is created, the target property may not even be installed yet on the target.  So instead we use a combination of targetPropertyName and target class and get the actual property when needed.`
```
Comment (kgrgreer) [inline src/foam/dao/Relationship.js]: multi-line string

File: src/foam/dao/Relationship.js [hunk 9]
```diff
@@ -520,18 +541,24 @@ foam.CLASS({
...
+  methods: [
+    function buildJavaClass(cls) {
+      this.SUPER(cls);
+      cls.method({
+        name: this.name,
+        visibility: 'public',
+        type: 'foam.dao.DAO',
+        args: [{ name: 'x', type: 'foam.core.X' }],
+        body: `return new foam.dao.RelationshipDAO.Builder(x)
+        .setSourceId(getId())
```
Comment (kgrgreer) [inline src/foam/dao/Relationship.js]: I think the next four lines should be indented.

File: src/foam/dao/Relationship.js [hunk 10]
```diff
@@ -572,9 +600,9 @@ foam.CLASS({
       factory: function() {
         var sourceProperty = this.sourceProperty;
         var targetProperty = this.targetProperty;
-        var targetDAOKey   = this.targetDAOKey;
+        var targetDAOKey = this.targetDAOKey;
```
Comment (kgrgreer) [inline src/foam/dao/Relationship.js]: Why mess-up alignment?

File: src/foam/dao/Relationship.js [hunk 11]
```diff
@@ -587,41 +615,55 @@ foam.CLASS({
...
   methods: [
+    function buildJavaClass(cls) {
+      this.SUPER(cls);
+      cls.method({
+        name: this.name,
+        visibility: 'public',
+        type: 'foam.dao.ManyToManyRelationship',
+        args: [{ name: 'x', type: 'foam.core.X' }],
+        body: `return new foam.dao.ManyToManyRelationshipImpl.Builder(x)
+        .setSourceId(getId())
```
Comment (kgrgreer) [inline src/foam/dao/Relationship.js]: Indent code.

File: src/foam/dao/Relationship.js (hunk 5 above)
Comment (shreymahendru) [reply src/foam/dao/Relationship.js]: All the changes I made to the formatting was to satisfy the linter which is based on the rules defined by us. I can change it back.

File: src/foam/dao/Relationship.js (hunk 6 above)
Comment (shreymahendru) [reply src/foam/dao/Relationship.js]: It was greater than 80 chars (which was one of our rules). But yea I'll change it back.

**Pull request merged**